lib/task_helpers/imports/custom_buttons.rb

```diff
@@ -0,0 +1,85 @@
+require "English"
+
+module TaskHelpers
+  class Imports
+    class CustomButtons
+      class ImportArInstances
+        def self.import(obj_hash)
+          new.import(obj_hash)
+        end
+
+        def import(obj_hash)
+          ActiveRecord::Base.transaction { obj_hash.each { |obj_def| create_object(*obj_def) } }
+        end
+
+        def create_object(class_name, obj_array)
+          klass = class_name.camelize.constantize
+
+          obj_array.collect do |obj|
+            begin
+              klass.create!(obj['attributes'].except('guid')).tap do |new_obj|
+                add_children(obj, new_obj)
+                add_associations(obj, new_obj)
+                try("#{class_name}_post", new_obj)
+              end
+            rescue StandardError
+              $log.send(:info, "Failed to create new instance [#{class_name}] with attributes #{obj['attributes'].inspect}")
+              $log.send(:info, "#{$ERROR_INFO} at #{$ERROR_POSITION}")
+              raise
+            end
+          end
+        end
+
+        def add_children(obj, new_obj)
+          if obj['children'].present?
+            obj['children'].each do |child|
+              new_obj.add_members(create_object(*child))
+            end
+          end
+        end
+
+        def add_associations(obj, new_obj)
+          if obj['associations'].present?
+            obj['associations'].each do |assoc|
+              new_obj.send("#{assoc.first}=", create_object(*assoc).first)
+            end
+          end
+        end
+
+        def custom_button_set_post(new_obj)
+          new_obj.set_data[:button_order] = new_obj.custom_buttons.collect(&:id)
+          new_obj.save!
+        end
+
+        def custom_button_post(new_obj)
+          check_user(new_obj)
+        end
+
+        def check_user(new_obj)
+          existing_user = User.find_by(:name => new_obj[:userid])
+          new_obj.update_attributes(:userid => existing_user.nil? ? "admin" : existing_user)
+        end
+      end
+
+      def import(options)
+        return unless options[:source]
+
+        glob = File.file?(options[:source]) ? options[:source] : "#{options[:source]}/CustomButtons.yaml"
+        Dir.glob(glob) do |filename|
+          begin
+            import_custom_buttons(filename)
+          rescue StandardError
+            warn("Error importing #{options[:source]} due to #{$ERROR_INFO} at #{$ERROR_POSITION}")
```

Comment:
Two things here (I think?), first is that, if we get here, aren't we getting here because of the `raise` from here: https://github.com/ManageIQ/manageiq/pull/17726/files#diff-a0ce4acbb7481b72b585132391c2dd4cR28, which means that this will just say something like "error because of raise at line 28?" and not give us actual good info?

Second is that I think there is the potential for `YAML.load_file` to fail if the YAML is messed up, in which case I don't remember if StandardError will catch it, cause it's a Psych error or something I think. Maybe?

In either case, maybe some specs ensuring the errors are doing what you want them to do are a good solution to my worries 😄